### ISSUE TITLE
Make quarkus.x.health.enabled configurable at run time

### ISSUE THREAD

**Issue by vsevel** [quarkusio/quarkus#22896]
### Description

There are some debates as to whether the readiness check should include checks on external systems.
We do not intend to do so for the "hundreds" ;-) of quarkus applications we are going to create.
We would like to make this a standard for all of these applications (and configure a standard env variable to disable those DS checks), and not rely on each app to specify `quarkus.datasource.health.enabled`
Unfortunately those properties are build time. We would like it to be run time instead, so that we can set the value using our deployment tool for all apps.

see also this [discussion](https://groups.google.com/g/quarkus-dev/c/pj6L7kzbIno/m/p5LEgE-gAAAJ)

### Implementation ideas

make `quarkus.x.health.enabled` run time

if not feasible, is there a way to set a property at build time from an extension?

**Comment by xstefank** [comment]
Hi @vsevel. Just letting you know that I am looking into this and evaluating how it could be done. In Quarkus itself, `quarkus.x.health.enabled` are used to produce or not to produce health check beans so this is why it is a build time property.

**Comment by xstefank** [comment]
So actually this will be possible with the new feature we will be introducing in SR Health 3.2.0 that allows disabling individual health checks - https://github.com/smallrye/smallrye-health/issues/128. Would this work for you? Do you know the full package name of the health checks coming from extensions?

**Comment by jmartisk** [comment]
Keep in mind mostly that if this is runtime, we will have to produce **all** relevant beans for extension health checks during build, because we won't know whether they will be used in the end. So applications will bloat more than necessary, based on how many such checks there will be.

**Comment by xstefank** [comment]
@jmartisk correct, but I don't see how we can get around that if the disabling should be dynamic at runtime.

**Comment by vsevel** [comment]
I understand. Is there a way to set a property at build time from an extension? i.e. using a build item.
we have developed an internal extension that all our quarkus applications will include. since disabling those healthchecks is going to be company wide, one option is to force disabling those health checks at build time through our extension.
If this is not possible, setting it at run time, and incurring a bigger built application will be acceptable. what are we talking about? a bigger native application? more memory consumed at run time? more time to boot the app? more time to execute a unit test? ...

we are specifically interested in disabling the datasource and vault health checks. but we will want to extend this to any check that has a dependency on an external system (message brokers, other types of dbs, ...)

**Comment by jmartisk** [comment]
The reason why we have runtime properties is that then you can set them per-deployment without having to rebuild, meaning different deployments of the same archive can have different configuration. Your reason seems to differ slightly ;) can't say how much of a performance issue it would bring, but it would affect most of Quarkus users, that's the main problem here.

I struggle to see why it's a problem to set a property for each project during build, but it's not a problem to set it during deployment. If you're going to change the pom.xml of all projects by adding the internal extension, why can't you at the same time add a line to `application.properties`?

Setting a build-time prop in an extension, to be honest I don't know, I suppose there's no "official" way, perhaps you could hack it in somehow.

**Comment by xstefank** [comment]
There is `SystemPropertyBuildItem` that you can use I believe. But with build time health disable we will not package health check class in your application so you won't have access to these health checks at runtime. Usually, I might be wrong here as I haven't checked all extensions but I can see only 1,2, or 3 HealthBuildItems being produced in one extension so that should not be that much of overhead and it would allow you to use these health checks at runtime and optionally disable them at runtime by explicitly stating which HealthCheck implementations should not be included in the responses with config property.

**Comment by jmartisk** [comment]
> There is `SystemPropertyBuildItem` that you can use I believe.

The javadoc of that says that it's "system property that will be set immediately on application startup" so processors (build-time) won't see it

**Comment by xstefank** [comment]
@jmartisk ok, thanks. So this won't work. But as you said, application.properties during build should.

**Comment by vsevel** [comment]
having a property at run time to disable specific health checks is sufficient indeed. I am not worried about the extra classes that will be bundled at build time in the app.

> I struggle to see why it's a problem to set a property for each project during build, but it's not a problem to set it during deployment.

we want to enforce that all projects do not activate certain checks, and not rely on people doing the right thing in their src/main/resources/application.properties. 
in our k8s deployment, we provision for them the necessary environment variables to enforce those company standards.

another solution would be to develop an enforcer, to make the build fail if we do not see that the check has been disabled.

the issue with the run time property, is that I suppose I would have to wait for this property to be supported in the different health checks I want to be able to deactivate? and every new extension that comes round, I may have to ask for the run time property to be supported as well.

so basically I see 3 solutions:
- wait for the run time property to be supported in all health checks I want to deactivate
- rely on the app doing it at build time, and develop an enforcer to make sure it is being done effectively
- find a way, if any, for our extension to produce a build time property

any other idea?

**Comment by jmartisk** [comment]
> wait for the run time property to be supported in all health checks I want to deactivate

Yeah so this is problematic because of the bloating, and also because all applicable Quarkus extensions would need to comply with this, which would mean a lot of work to be coordinated.

> rely on the app doing it at build time, and develop an enforcer to make sure it is being done effectively

This could do it. Perhaps your extension could add some logic that verifies this during boot time, and throws an exception in case if that check fails. It could maybe iterate through the list of active health checks, and assert that some specific checks aren't there.

> find a way, if any, for our extension to produce a build time property

Potentially this could work, but I'm not completely sure. And it would probably be an ugly hack that might stop working if something changes in Quarkus.

So I would personally recommend the second option.

**Comment by vsevel** [comment]
so wait. I missed what you were doing in https://github.com/smallrye/smallrye-health/pull/357/. this will allow to deactivate any check at run time using property `"io.smallrye.health.check." + checkClassName + ".enabled"`? 
and the only inconvenience is a bit of bloating?
if so, I would go for that.
when is this going to be available in quarkus?

**Comment by xstefank** [comment]
The new feature will be in the SR Health and will only require the full package name of the health check implementation so it won't be dependent on any extension specifics. You just need to know the name of the health check and then you can use `io.smallrye.health.check.<health-check-class-name>.enabled=false`.

**Comment by vsevel** [comment]
perfect. 
do you have a timeline/release for SR Health and its inclusion in Quarkus?

**Comment by xstefank** [comment]
hopefully by the end of the week :)

**Comment by vsevel** [comment]
ok. I will be following this. thanks for looking into it. keep up the good work!

**Comment by jmartisk** [comment]
Ah ok that actually seems it could do it :) Just please bear in mind that 2.7.0.CR1 is planned for today, which might complicate the upgrade a little if it's to be present in 2.7 (I think 2.7 will have its own branch soon, so a backport will be needed)

**Comment by xstefank** [comment]
@jmartisk @vsevel I wasn't following this lately. Do I need to backport upgrade somewhere?

**Comment by vsevel** [comment]
hi @xstefank 
I assumed https://github.com/smallrye/smallrye-health/pull/357 was in 2.7.0
it is not?

**Comment by vsevel** [comment]
ok. no problem. we can wait. thanks for asking.

**Comment by vsevel** [comment]
hi @xstefank 
I have moved to `2.8.0` and I have added `quarkus.datasource.health.enabled=false` to my config.
I am still seeing a `Database connections health check` on `q/health/ready`:

```
curl -k https://myhost/q/health/ready

{
    "status": "UP",
    "checks": [
        {
            "name": "Database connections health check",
            "status": "UP"
        }
    ]
}
```

I was expecting the check to not be there. am I missing anything?

**Comment by vsevel** [comment]
hmm. got it. it is `io.smallrye.health.check.io.quarkus.agroal.runtime.health.DataSourceHealthCheck.enabled=false`
not sure why I was thinking `quarkus.datasource.health.enabled` could be used??

**Comment by xstefank** [comment]
Hi @vsevel, it can actually be also `quarkus.smallrye-health.check."io.quarkus.agroal.runtime.health.DataSourceHealthCheck".enabled=false` for Quarkus specific way.

The reason why `quarkus.datasource.health.enabled` doesn't work runtime is that this property is actually telling compile time whether we will include checks meaning check classes in runtime or not. For runtime disable of individual checks you need to use the above properties. But the check classes are included in runtime just not processed.

**Comment by vsevel** [comment]
works like a charm @xstefank thanks!
